heft-plugins/heft-webpack5-plugin/src/WebpackPlugin.ts

```diff
@@ -0,0 +1,270 @@
+          { name: PLUGIN_NAME, stage: Number.MIN_SAFE_INTEGER },
+          (webpackConfiguration: unknown) => {
+            const webpackVersions: IWebpackVersions = getWebpackVersions();
+            bundle.properties.webpackVersion = webpack.version;
+            bundle.properties.webpackDevServerVersion = webpackVersions.webpackDevServerVersion;
+
+            return webpackConfiguration;
+          }
+        );
+
+        bundle.hooks.configureWebpack.tapPromise(PLUGIN_NAME, async (existingConfiguration: unknown) => {
+          const logger: ScopedLogger = heftSession.requestScopedLogger('configure-webpack');
+          if (existingConfiguration) {
+            logger.terminal.writeVerboseLine(
+              'Skipping loading webpack config file because the webpack config has already been set.'
+            );
+            return existingConfiguration;
+          } else {
+            return await WebpackConfigurationLoader.tryLoadWebpackConfigAsync(
+              logger,
+              heftConfiguration.buildFolder,
+              build.properties
+            );
+          }
+        });
+
+        bundle.hooks.run.tapPromise(PLUGIN_NAME, async () => {
+          await this._runWebpackAsync(
+            heftSession,
+            heftConfiguration,
+            bundle.properties as IWebpackBundleSubstageProperties,
+            build.properties,
+            heftConfiguration.terminalProvider.supportsColor
+          );
+        });
+      });
+    });
+  }
+
+  private async _runWebpackAsync(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    bundleSubstageProperties: IWebpackBundleSubstageProperties,
+    buildProperties: IBuildStageProperties,
+    supportsColor: boolean
+  ): Promise<void> {
+    const webpackConfiguration: IWebpackConfiguration = bundleSubstageProperties.webpackConfiguration;
+    if (!webpackConfiguration) {
+      return;
+    }
+
+    const logger: ScopedLogger = heftSession.requestScopedLogger('webpack');
+    const webpackVersions: IWebpackVersions = getWebpackVersions();
+    if (bundleSubstageProperties.webpackVersion !== webpackVersions.webpackVersion) {
+      logger.emitError(
+        new Error(
+          `The Webpack plugin expected to be configured with Webpack version ${webpackVersions.webpackVersion}, ` +
+            `but the configuration specifies version ${bundleSubstageProperties.webpackVersion}. ` +
+            'Are multiple versions of the Webpack plugin present?'
+        )
+      );
+    }
+
+    if (bundleSubstageProperties.webpackDevServerVersion !== webpackVersions.webpackDevServerVersion) {
+      logger.emitError(
+        new Error(
+          `The Webpack plugin expected to be configured with webpack-dev-server version ${webpackVersions.webpackDevServerVersion}, ` +
+            `but the configuration specifies version ${bundleSubstageProperties.webpackDevServerVersion}. ` +
+            'Are multiple versions of the Webpack plugin present?'
```

Comment:
Multiple versions of the webpack plugin should be an allowed scenario and may well be something a developer would want to do, hence why "configureWebpack" should be a hook on the webpack plugin instance, or on an "IWebpackPluginHooks" object, obtained by `WebpackPlugin.getHooks(heftSession)`, not on the bundle substage.

Reply:
i'm curious what the use case is for multiple webpack versions?

Reply:
Backwards compatibility is the main reason I can think of. Building both so that you can compare the outputs between the two side by side, especially during upgrade.

Reply:
very interesting. i would have never considered :) perhaps i should have during this last upgrade haha